File: data_structures/01_LRU_cache/test_lru_cache.cpp
#include "lru_cache.cpp"
#include <gtest/gtest.h>
#include <vector>
#include <chrono>
#include <iostream>
 
using namespace std;


// A simple struct to hold test results
struct TestResult {
    string test_name;
    chrono::duration<double> elapsed_time;
};

// Custom listener
class CustomTestListener : public ::testing::EmptyTestEventListener {
private:
    vector<TestResult> results;
    chrono::time_point<chrono::high_resolution_clock> start_time;

public:
    // Called before a test starts
    void OnTestStart(const ::testing::TestInfo& test_info) override {
        start_time = chrono::high_resolution_clock::now();
    }

    // Called after a test ends
    void OnTestEnd(const ::testing::TestInfo& test_info) override {
        auto end_time = chrono::high_resolution_clock::now();
        chrono::duration<double> elapsed_time = end_time - start_time;
        results.push_back({test_info.name(), elapsed_time});
    }

    // Called after all tests have run
    void OnTestProgramEnd(const ::testing::UnitTest& unit_test) override {
        // Now print all the results
        cout << "\nTest Time Results:\n\n";
        int nameWidth = 60;
        for (const auto& result : results) {
            cout << left << setw(nameWidth) << result.test_name 
                 << "time: " << scientific << setprecision(2) 
                 << result.elapsed_time.count() << " s\n";
        }
        cout << endl;
    }
};


class TestLRUCache : public ::testing::Test {
protected:
    LRUCache* cache; // Each test will have its own pointer to LRUCache instance

    void SetUp() override {
        // Code here will be called immediately after the constructor 
        // (right before each test).
        cache = new LRUCache();
    }

    void TearDown() override {
        // Code here will be called immediately after each test 
        // (right before the destructor).
        delete cache;
    }
};


TEST_F(TestLRUCache, test1_Cache1) {
    cache->setCapacity(2);
    cache->put(1, 1);
    cache->put(2, 2);
    int result = cache->get(1);
    int expected = 1;
    ASSERT_EQ(expected, result);
    
    cache->put(3, 3);
    result = cache->get(2);
    expected = -1;
    ASSERT_EQ(expected, result);

    cache->put(4, 4);
    result = cache->get(1);
    expected = -1;
    ASSERT_EQ(expected, result);

    result = cache->get(3);
    expected = 3;
    ASSERT_EQ(expected, result);

    result = cache->get(4);
    expected = 4;
    ASSERT_EQ(expected, result);
}

TEST_F(TestLRUCache, test2_Cache2) {
    cache->setCapacity(2);
    cache->put(1, 1);
    cache->put(2, 2);
    cache->put(2, 3);
    int result = cache->get(2);
    int expected = 3;
    ASSERT_EQ(expected, result);
}


int main(int argc, char **argv) {
    ::testing::InitGoogleTest(&argc, argv);

    // Adds the custom listener to the end. Google Test takes the ownership.
    ::testing::TestEventListeners& listeners = ::testing::UnitTest::GetInstance()->listeners();
    listeners.Append(new CustomTestListener);

    return RUN_ALL_TESTS();
}

File: data_structures/01_LRU_cache/test_lru_cache.py
from unittest import TestCase, TestResult, TestLoader, TextTestRunner
from lru_cache import LRUCache
from time import time
import numpy as np

from functools import wraps


class CustomTestResult(TestResult):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.test_times = []

    def startTest(self, test):
        self._start_time = time()
        super().startTest(test)

    def stopTest(self, test):
        elapsed_time = time() - self._start_time
        self.test_times.append((test.id().split('.')[-1], elapsed_time))
        super().stopTest(test)

    def printTestTimes(self):
        name_width = 60
        print("\nTest Time Results:\n")
        for name, elapsed_time in self.test_times:
            print(f"{name:<{name_width}}time: {elapsed_time:.2e} s")
        print()


def time_tests(cls):
    """Class decorator for adding time measurements to all test methods of a TestCase."""
    for attr in dir(cls):
        if attr.startswith('test'):
            old_method = getattr(cls, attr)

            @wraps(old_method)
            def new_method(self, *args, method=old_method, **kwargs):
                method(self, *args, **kwargs)
            setattr(cls, attr, new_method)
    return cls


class TestLRUCache(TestCase):
    def setUp(self):
        # This method is called before each test
        self.cache = LRUCache()

    def test1_Cache1(self):
        self.cache.capacity = 2
        self.cache.put(1, 1)
        self.cache.put(2, 2)
        result = self.cache.get(1)
        expected = 1
        self.assertEqual(result, expected)

        self.cache.put(3, 3)
        result = self.cache.get(2)
        expected = -1
        self.assertEqual(result, expected)

        self.cache.put(4, 4)
        result = self.cache.get(1)
        expected = -1
        self.assertEqual(result, expected)

        result = self.cache.get(3)
        expected = 3
        self.assertEqual(result, expected)

        result = self.cache.get(4)
        expected = 4
        self.assertEqual(result, expected)

    def test2_Cache2(self):
        self.cache.capacity = 2
        self.cache.put(1, 1)
        self.cache.put(2, 2)
        self.cache.put(2, 3)
        result = self.cache.get(2)
        expected = 3
        self.assertEqual(result, expected)

    def tearDown(self):
        del self.cache  # Ensure the reference is deleted.


if __name__ == '__main__':
    suite = TestLoader().loadTestsFromTestCase(TestLRUCache)
    result = CustomTestResult()
    runner = TextTestRunner(verbosity=2)
    # runner.run(suite)
    suite.run(result)
    result.printTestTimes()
